### ISSUE TITLE
Large slowdown with map

### ISSUE THREAD

**Issue by zenna** [jrevels/Cassette.jl#98]
I don't know if this is because map is generally is harder to infer types? Either way there is significant slowdown (1000x) using Cassette in this simple example using map and a 2x slowdown with explicit iteration.

```julia
using Cassette
using BenchmarkTools

kse(d, α = 10000.0) = α * d

Cassette.@context AlphaCtx

function atα(α, f, args...)
  ctx = AlphaCtx(metadata = α)
  Cassette.overdub(ctx, f, args...)
end

Cassette.execute(ctx::AlphaCtx, ::typeof(kse), x, α) = kse(x, ctx.metadata)

function test1(x)
  kse(x)
end

function test2(x)
  atα(10000, map, kse, x)
end


function test3(x)
  x_ = similar(x)
  for i = 1:length(x)
    x_[i] = atα(10000, kse, x[i])
  end
  x_
end
```

Test benchmarking results 

```

julia> @benchmark test1(x) setup = (x = rand(10000))
BenchmarkTools.Trial: 
  memory estimate:  78.20 KiB
  allocs estimate:  2
  --------------
  minimum time:     4.548 μs (0.00% GC)
  median time:      5.770 μs (0.00% GC)
  mean time:        7.311 μs (15.93% GC)
  maximum time:     4.317 ms (99.54% GC)
  --------------
  samples:          10000
  evals/sample:     7


julia> @benchmark test2(x) setup = (x = rand(10000))
BenchmarkTools.Trial: 
  memory estimate:  1.14 MiB
  allocs estimate:  59497
  --------------
  minimum time:     4.050 ms (0.00% GC)
  median time:      4.280 ms (0.00% GC)
  mean time:        4.451 ms (1.86% GC)
  maximum time:     41.354 ms (89.14% GC)
  --------------
  samples:          1118
  evals/sample:     1

julia> @benchmark test3(x) setup = (x = rand(10000))
BenchmarkTools.Trial: 
  memory estimate:  78.20 KiB
  allocs estimate:  2
  --------------
  minimum time:     6.533 μs (0.00% GC)
  median time:      7.132 μs (0.00% GC)
  mean time:        9.825 μs (18.45% GC)
  maximum time:     6.680 ms (99.65% GC)
  --------------
  samples:          10000
  evals/sample:     5

```

**Comment by domluna** [comment]
@zenna 

Running this on Cassette master (v0.2) I get

```
julia> @benchmark test1(x) setup = (x = rand(10000))
BenchmarkTools.Trial:
  memory estimate:  78.20 KiB
  allocs estimate:  2
  --------------
  minimum time:     3.505 μs (0.00% GC)
  median time:      7.615 μs (0.00% GC)
  mean time:        10.564 μs (36.90% GC)
  maximum time:     520.787 μs (95.38% GC)
  --------------
  samples:          10000
  evals/sample:     7

julia> @benchmark test2(x) setup = (x = rand(10000))
BenchmarkTools.Trial:
  memory estimate:  78.22 KiB
  allocs estimate:  3
  --------------
  minimum time:     3.986 μs (0.00% GC)
  median time:      8.180 μs (0.00% GC)
  mean time:        10.949 μs (34.09% GC)
  maximum time:     368.636 μs (93.49% GC)
  --------------
  samples:          10000
  evals/sample:     7

julia> @benchmark test3(x) setup = (x = rand(10000))
BenchmarkTools.Trial:
  memory estimate:  78.20 KiB
  allocs estimate:  2
  --------------
  minimum time:     6.003 μs (0.00% GC)
  median time:      9.535 μs (0.00% GC)
  mean time:        12.374 μs (31.20% GC)
  maximum time:     370.725 μs (97.10% GC)
  --------------
  samples:          10000
  evals/sample:     5```

Change `execute` to `overdub` since it's deprecated now.

**Comment by jrevels** [comment]
AFAICT your OP is mostly fixed on Cassette master (i.e. will be fixed in Cassette v0.2 https://github.com/JuliaLang/METADATA.jl/pull/20164). Woohoo! If you test it on v0.2, remember to change `Cassette.execute` in your OP to `Cassette.overdub` (the former is now deprecated).  EDIT: looks like @domluna beat me to it

However, in the OP, `overdub` vs. no `overdub` + context creation overhead + `map` vs. no `map` are all being tested together in the same benchmark comparison (`test1` vs `test2`). `test3` doesn't seem to have a non-Cassette version to compare against at all, so little information can be inferred there.

Just to do due diligence, here are some benchmarks to do some more direct apples-to-apples comparisons:

```julia
using Cassette, BenchmarkTools

Cassette.@context AlphaCtx

kse(d, α = 10000.0) = α * d

Cassette.overdub(ctx::AlphaCtx, ::typeof(kse), x, α) = kse(x, ctx.metadata)

test_map(x) = map(kse, x)

test_map_overdub(ctx, x) = Cassette.overdub(ctx, test_map, x)

function test_iteration(x)
    x_ = similar(x)
    for i = 1:length(x)
        x_[i] = kse(x[i])
    end
    x_
end

test_iteration_overdub(ctx, x) = Cassette.overdub(ctx, test_iteration, x)

atα(α, args...) = Cassette.overdub(AlphaCtx(metadata = α), args...)

function test_iteration_overdub_atα(x)
    x_ = similar(x)
    for i = 1:length(x)
        x_[i] = atα(10000, kse, x[i])
    end
    x_
end
```

So we can compare:

-  `test_map` vs. `test_map_overdub` to see if there's any overdubbing overhead for `map(kse, x)`
- `test_iteration` vs. `test_iteration_overdub` to see if there's any overdubbing overhead for manual iteration
- `test_iteration_overdub` vs. `test_iteration_overdub_atα` to see if there's any overhead for context instantiation + overdubbing only per-element instead of the whole function

Results:

```
julia> versioninfo()
Julia Version 1.0.2
Commit d789231e99 (2018-11-08 20:11 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin17.5.0)
  CPU: Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)

julia> @benchmark test_map(x) setup=(x=rand(100000))
BenchmarkTools.Trial:
  memory estimate:  781.34 KiB
  allocs estimate:  3
  --------------
  minimum time:     64.745 μs (0.00% GC)
  median time:      94.349 μs (0.00% GC)
  mean time:        240.873 μs (11.51% GC)
  maximum time:     39.521 ms (98.80% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark test_map_overdub(ctx, x) setup=(x=rand(100000); ctx=AlphaCtx(metadata=rand(Int)))
BenchmarkTools.Trial:
  memory estimate:  781.34 KiB
  allocs estimate:  3
  --------------
  minimum time:     63.961 μs (0.00% GC)
  median time:      95.011 μs (0.00% GC)
  mean time:        240.702 μs (11.35% GC)
  maximum time:     42.094 ms (98.85% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark test_iteration(x) setup=(x=rand(100000))
BenchmarkTools.Trial:
  memory estimate:  781.33 KiB
  allocs estimate:  2
  --------------
  minimum time:     72.868 μs (0.00% GC)
  median time:      116.137 μs (0.00% GC)
  mean time:        261.518 μs (9.15% GC)
  maximum time:     36.271 ms (98.64% GC)
  --------------
  samples:          9768
  evals/sample:     1

julia> @benchmark test_iteration_overdub(ctx, x) setup=(x=rand(100000); ctx=AlphaCtx(metadata=rand(Int)))
BenchmarkTools.Trial:
  memory estimate:  781.33 KiB
  allocs estimate:  2
  --------------
  minimum time:     72.817 μs (0.00% GC)
  median time:      110.987 μs (0.00% GC)
  mean time:        264.399 μs (9.99% GC)
  maximum time:     41.803 ms (98.66% GC)
  --------------
  samples:          9656
  evals/sample:     1

# measure the effects of context creation overhead vs. previous `test_iteration_*` (seems to be negligble)
julia> @benchmark test_iteration_overdub_atα(x) setup=(x=rand(100000))
  BenchmarkTools.Trial:
    memory estimate:  781.33 KiB
    allocs estimate:  2
    --------------
    minimum time:     72.932 μs (0.00% GC)
    median time:      106.930 μs (0.00% GC)
    mean time:        266.482 μs (9.79% GC)
    maximum time:     45.649 ms (98.81% GC)
    --------------
    samples:          9585
    evals/sample:     1
```

**Comment by zenna** [comment]
Cool.

Updating to v0.2 fixed the example as I wrote it.  This was supposed to be a minimum working example, but in my actual code I still see a big slowdown there.   Actually the only difference between the MWE and the real code is that kse isn't called directly but nested much deeper in the call stack.  I'll try to construct another MWE and post another issue.
